sql/parser: column type refactoring / simplification

knz opened the pull request:
This patchset extracts the ColumnType definitions away from the `parser` package and thereby simplifies their names.

For example:

| Before | After |
|-----------|--------|
| parser.IntColType | coltypes.TInt |
| parser.IntColTypeInt | coltypes.Int |
| parser.BoolColType | coltypes.TBool |
| parser.BoolColTypeBool | coltypes.Bool |

The names starting with `T` are the Go types used to encode the column types, whereas the names without a `T` prefix are the immutable instances that implement the ColumnType interface.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19973)
<!-- Reviewable:end -->

knz commented:
cc @jordanlewis

knz committed:
sql/parser: break the dependency col_types.go -> format.go

justinj commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> love the work you've been putting into these changes!

---

Review status: 0 of 15 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19973#-:-KybSW_MBLveon9AXEdi:b-uf5990)*
<!-- Sent from Reviewable.io -->

knz committed:
sql/parser: new method (*NumVal).AsInt32()

knz committed:
sql/parser: break the dependency ColumnType -> Expr

The array column type has constant bounds, if any. The PG grammar only
accepts literals as bounds when using array types.

This patch acknowledges this, makes the bound field of ArrayColType an
`[]int32`, and thereby removes the conversion overhead when
constructing table descriptors.

knz committed:
sql/parser: move the ColumnType constructors and pre-allocs to their own file

knz committed:
sql/parser: move the ColumnType converters to their own file

knz committed:
sql/parser: move ColumnType definitions into separate files

knz committed:
sql/parser: move the ColumnType definitions to a new package sql/coltypes

knz committed:
sql/parser: move LimitDecimalWidth to decimal.go where it belongs

knz committed:
sql/parser: move canBeInArrayColType to col_types_arrays.go where it belongs

knz commented:
TFYR!

knz committed:
sql/parser: remove the coltypes aliases

knz committed:
sql/coltypes: remove the stub aliases

knz committed:
sql/coltypes: destutter the column type names.

Summary:

- `coltypes.ColumnType` -> `coltypes.T`
- `coltypes.IntColType` -> `coltypes.TInt`, `FloatColType` -> `TFloat`, etc.
- `coltypes.IntColTypeInt` -> `coltypes.Int`, `coltypes.BoolColTypeBool` -> `coltypes.Bool`, etc.
